### ISSUE TITLE
Allow building an ApolloClient on top of another one changing the underlying OkHttpClient

### ISSUE THREAD

**Issue by francescocervone** [apollographql/apollo-kotlin#3759]
**Use case**
Let's say I have 
```kotlin
val okHttpClient1: OkHttpClient = someFactory()
val client1: ApolloClient = ApolloClient.Builder()
  .okHttpClient(okHttpClient1)
  .build()
```

Then I need to create a new Apollo client on top of `client1` passing a new `OkHttpClient`

```kotlin
val okHttpClient2: OkHttpClient = someNewFactory()
val client2: ApolloClient = client1.newBuilder()
  .okHttpClient(okHttpClient2)
  .build()
```

With Apollo 3.0.0 this is not possible, since `build` throws an exception
```
Apollo: 'httpEngine' has no effect if 'networkTransport' is set
```

I cannot even put `networkTransport` to `null` because `ApolloClient.Builder.networkTransport(...)` expects a non-nullable variable.

**Describe the solution you'd like**
I expect that Apollo allows me to build an `ApolloClient` on top of another one (built with an `OkHttpClient`) with a new `OkHttpClient`.

**Comment by martinbonnin** [comment]
Good catch, that's a side effect of `ApolloClient.Builder.okHttpClient()` being a "shortcut" to simplify the setup. 

You can use `ApolloClient.Builder.networkTransport()` to override it:

```kotlin
    apolloClient
        .newBuilder()
        .networkTransport(
            HttpNetworkTransport.Builder()
                .serverUrl(serverUrl2)
                .okHttpClient(okHttpClient2)
                .build()
        )
```

I think making `networkTransport` nullable in `ApolloClient.Builder.networkTransport()` is a reasonable thing too to reset the network transport.

The other solution to respect 100% the `Builder` semantics is to store all "shortcut" properties in `ApolloClient` and forward them in `newBuilder()`. No strong opinion on which one is better yet. That last one is more consistent I believe but has a small price in terms of making `ApolloClient` bigger and more complex. Any thoughts?

**Comment by francescocervone** [comment]
I think making `networkTransport` nullable is a better fit.
Since `ApolloClient.Builder.okHttpClient` is an extension function hiding the detail about of how Apollo actually works with `OkHttpClient`, it should also hide the detail that the `networkTransport` (and maybe `subscriptionNetworkTransport`) must be set to null again.

I expect something like this:
```kotlin
fun ApolloClient.Builder.okHttpClient(okHttpClient: OkHttpClient) = apply {
  networkTransport(null)
  subscriptionNetworkTransport(null) // I'm pretty sure this is also needed
  httpEngine(DefaultHttpEngine(okHttpClient))
  webSocketEngine(DefaultWebSocketEngine(okHttpClient))
}
```

In general I don't see any problems in having nullable parameters in builders. They allow the client to reset parts of the configuration.

I don't know about storing "shortcut" properties, I'm not sure I understand: what are shortcut properties? The `OkHttpClient`?

**Comment by martinbonnin** [comment]
> I don't know about storing "shortcut" properties, I'm not sure I understand: what are shortcut properties? The OkHttpClient?

Exactly! `ApolloClient.Builder` contains a `httpEngine` var that is not in `ApolloClient` because it's only used to build the `networkTransport` but that duplicates some information. See [the logic that builds `ApolloClient`](https://github.com/apollographql/apollo-kotlin/blob/5988505e6c590538b6f989f369f5f086b4b6139f/apollo-runtime/src/commonMain/kotlin/com/apollographql/apollo3/ApolloClient.kt#L445):

```kotlin
    fun build(): ApolloClient {
      val networkTransport = if (_networkTransport != null) {
        check(httpServerUrl == null) {
          "Apollo: 'httpServerUrl' has no effect if 'networkTransport' is set"
        }
        check(httpEngine == null) {
          "Apollo: 'httpEngine' has no effect if 'networkTransport' is set"
        }
        check(httpInterceptors.isEmpty()) {
          "Apollo: 'addHttpInterceptor' has no effect if 'networkTransport' is set"
        }
        check(httpExposeErrorBody == null) {
          "Apollo: 'httpExposeErrorBody' has no effect if 'networkTransport' is set"
        }
        _networkTransport!!
      } else {
        check(httpServerUrl != null) {
          "Apollo: 'serverUrl' is required"
        }
        HttpNetworkTransport.Builder()
            .serverUrl(httpServerUrl!!)
            .apply {
              if (httpEngine != null) {
                httpEngine(httpEngine!!)
              }
              if (httpExposeErrorBody != null) {
                exposeErrorBody(httpExposeErrorBody!!)
              }
            }
            .interceptors(httpInterceptors)
            .build()
      }
```

As soon as `httpEngine` (which itself contains the okHttpClient) is used to create a `NetworkTransport`, some information is lost. So we _could_ forward the `httpEngine` val to `ApolloClient` so that when calling `newBuilder()`, the information that the `NetworkTransport` came from a `okHttpClient()` call is still there. It's more boilerplate but that'd make the `ApolloClient.Builder` behave more like a real builder.

**Comment by francescocervone** [comment]
If I am not mistaken, I suppose you have to options here:
1) You move the `networkTransport/subscriptionNetworkTransport` creation inside `ApolloClient`. But then you have some variables only needed to create a new builder. In order to make the builder more consistent, you are making the implementation of `ApolloClient` just the opposite and more error prone. Moreover, you are preventing clients to build their own `networkTransport`, but maybe that's just ok because they can build it in a "destructured" way, passing in url, engine and interceptors.
2) You keep the `networkTransport/subscriptionNetworkTransport` in the builder API forwarding everything to the `ApolloClient`. This way you are just moving the decision of "who wins" among transport and engine. Moreover you are allowing clients to mess up with those values (I can pass a `networkTransport` and then pass a `httpEngine`).

I kind of used this message as talking with my rubber duck 😄, and maybe solution number 1 (removing transport from the builder) is a good balance.

**Comment by BoD** [comment]
A fix has just been merged, and will be available in the next release. In the meantime it you want to have a look, you can use the [snapshots](https://github.com/apollographql/apollo-kotlin#snapshots).

**Comment by martinbonnin** [comment]
Fixed in 3.1.0
